src/main/java/gregtech/common/metatileentities/storage/CachedRecipeData.java

```diff
@@ -74,17 +78,47 @@ public boolean getIngredientEquivalent(int slot) {
             return true;
         }
 
+        ItemStack previousStack = recipe.getCraftingResult(inventory);
+
+        Map<ItemStackKey, Boolean> map = replaceAttemptMap.computeIfAbsent(slot, (m) -> new Object2BooleanOpenHashMap<>());
+
         //iterate stored items to find equivalent
         for (ItemStackKey itemStackKey : itemSources.getStoredItems()) {
+            if (map.containsKey(itemStackKey)) {
+                if (!map.get(itemStackKey)) {
+                    continue;
+                } else {
+                    return true;
+                }
+            }
+
             ItemStack itemStack = itemStackKey.getItemStack();
+
+            boolean matched = false;
+            //Matching shapeless recipes actually is very bad for performance, as it checks the entire
+            //recipe ingredients recursively, so we fail early here if none of the recipes ingredients can
+            //take the stack
+            for (Ingredient in : recipe.getIngredients()) {
+                if (in.apply(itemStack)) {
+                    matched = true;
+                    break;
+                }
+            }
+            if (!matched) {
+                map.put(itemStackKey, false);
+                continue;
+            }
+
             //update item in slot, and check that recipe matches and output item is equal to the expected one
             inventory.setInventorySlotContents(slot, itemStack);
-            if (recipe.matches(inventory, itemSources.getWorld())) {
+            if (recipe.matches(inventory, itemSources.getWorld()) && ItemStack.areItemStacksEqual(recipe.getCraftingResult(inventory), previousStack)) {
```

Comment:
Do we want to check the Equality before checking `matches`, so that we might be able to avoid more matches calls?

Reply:
theres only one IRecipe instance here; the getCraftingResult method should also return a copy of the itemstack in the output